src/nvim/eval.c

```diff
@@ -10681,29 +10691,31 @@ static void f_jobclose(typval_T *argvars, typval_T *rettv)
     return;
   }
 
-  Job *job = job_find(argvars[0].vval.v_number);
-
-  if (!is_user_job(job)) {
-    // Invalid job id
+  TerminalJobData *data = pmap_get(uint64_t)(jobs, argvars[0].vval.v_number);
+  if (!data) {
     EMSG(_(e_invjob));
     return;
   }
 
+  Process *proc = (Process *)&data->proc;
+
   if (argvars[1].v_type == VAR_STRING) {
     char *stream = (char *)argvars[1].vval.v_string;
     if (!strcmp(stream, "stdin")) {
-      job_close_in(job);
-      ((TerminalJobData *)job_data(job))->stdin_closed = true;
+      stream_close(proc->in, NULL);
```

Comment:
`jobclose()` could be called for the same stream multiple times. `stream_close()` would `assert()`, but `process_close_{in,out,err,streams}` should be safe to use here and in the following cases.